### ISSUE TITLE
HTTP3/QUIC support

### ISSUE THREAD

**Issue by TheLastGimbus** [home-assistant/android#1390]
I just recently switched from Nginx to Caddy for reverse proxy, and OH MY GOD IT'S SO GOOD

Caddy makes it super easy to add experimental HTTP3 protocol, which uses [quic](https://blog.cloudflare.com/the-road-to-quic/), which makes it *much faster :rocket:*

Is there any chance to add http3 support for Home Assistant app?

**Comment by dshokouhi** [comment]
not sure on the specifics but wouldnt HA core and the frontend need to support this first?

**Comment by TheLastGimbus** [comment]
I'm not sure, honestly...

As far as I understand, if you run behind a reverse-proxy, you can customise any HTTP stuff directly from that proxy - add TLS, add headers, etc

And since connection between proxy and hass is internal (0ms), it doesn't matter that hass itself doesn't have http3

(But I may be wrong...)

**Comment by TheLastGimbus** [comment]
Yes, after a lot of messing, I succesfully got some HTTP/3 connections, on login page for many parts:

![image](https://user-images.githubusercontent.com/40139196/110328180-47ae2700-801b-11eb-8e45-0b1393901213.png)
![image](https://user-images.githubusercontent.com/40139196/110328350-8ba12c00-801b-11eb-853d-201350a990c8.png)


For dashboard tho, it seems to catch up less often, altough *it works*:

![image](https://user-images.githubusercontent.com/40139196/110328306-7af0b600-801b-11eb-8668-d6d1343c650a.png)

Just thanks to Caddy, no hass core support required :100:

**Comment by TheLastGimbus** [comment]
Update: it doesn't seem to fully work :sweat_smile: I don't know if it's fault of Caddy, or Firefox, or my router or what, but lovelace can't load when http3 is enabled in Caddy - disabling it makes it work again

That just gets me on the ground that quic isn't ready yet :( But maybe official support in app would make it better :+1:

**Comment by TheLastGimbus** [comment]
Okhttp doesn't have the support *yet* either: square/okhttp#907

**Comment by stonith** [comment]
> 

Curious what doesn't work properly with http3 enabled on your reverse proxy?

**Comment by TheLastGimbus** [comment]
I recently re-installed my Rpi from ubuntu to hass os, set up this caddy plugin: https://github.com/einschmidt/addon-caddy-2

And currently, nothing works:

![image](https://github.com/home-assistant/android/assets/40139196/e2d2bcd8-1302-4685-a1b7-59d14af0c7ea)

Idk if hass os is blocking udp or something, but right now even http2 doesnt work :D

**Comment by stonith** [comment]
I recently setup envoy in front of HA and the iOS app uses http2 mostly and sometimes http3. Safari on osx and iOS don't seem to use http3 consistently though, chrome will use http3 much more consistently. I haven't tried Android but I'd assume it's more consistent with http3.

**Comment by Gunni** [comment]
I have been running caddy for years in front of all sorts of things, including HA, but recently i have started trying to push clients via my proxy to upgrade to using h3 with a fall back to h2.

1. First I get alt-svc for free via caddy, yay, some things pick up h2, but almost never h3, and websocket, no
    - [RFC 8441](https://datatracker.ietf.org/doc/rfc8441/) - Bootstrapping WebSockets with HTTP/2
    - [RFC 9220](https://datatracker.ietf.org/doc/rfc9220/) - Bootstrapping WebSockets with HTTP/3
3. Then I added HTTPS-RR, almost no clients use that afaict
4. Then recently I tried adding these headers towards clients:
    - ```caddyfile
      @is_h1 expression {http.request.proto}.startsWith("HTTP/1")
      # If protocol is HTTP/1.x → advertise both h3 and h2
      header @is_h1 {
            Alt-Svc "h3=\":443\"; ma=2147483647; persist=1, h2=\":443\"; ma=2147483647; persist=1"
            # Disabled for others to stop torturing HA app
            #Upgrade "HTTP/3.0, HTTP/2.0"
            Connection "upgrade"
      }

      @is_h2 expression {http.request.proto}.startsWith("HTTP/2")
      # If protocol is HTTP/2 → advertise only h3
      header @is_h2 {
            Alt-Svc "h3=\":443\"; ma=2147483647; persist=1"
      }
      ```

But notice that commented out line, if I send `Upgrade "HTTP/3.0, HTTP/2.0"` to the HA happ, it does kinda work, but the app constantly disconnects and needs to reconnect and if left open and then do an action it gets an error, bah...

I know I'm doing abnormal things, but I just want clients to understand that they should get off [HTTP 1.1](https://http1mustdie.com/)...
